src/Mvc/Mvc.Testing/src/WebApplicationFactory.cs

```diff
@@ -512,22 +513,53 @@ protected virtual void Dispose(bool disposing)
 
             if (disposing)
             {
-                foreach (var client in _clients)
+                if (!_disposedAsync)
                 {
-                    client.Dispose();
+                    DisposeAsync()
+                        .AsTask()
+                        .ConfigureAwait(false)
+                        .GetAwaiter()
+                        .GetResult();
                 }
+            }
 
-                foreach (var factory in _derivedFactories)
-                {
-                    factory.Dispose();
-                }
+            _disposed = true;
+        }
+
+        /// <inheritdoc />
+        public async ValueTask DisposeAsync()
+        {
+            if (_disposed)
+            {
+                return;
+            }
+
+            if (_disposedAsync)
+            {
+                return;
+            }
 
-                _server?.Dispose();
-                _host?.StopAsync().Wait();
+            foreach (var client in _clients)
+            {
+                client.Dispose();
+            }
+
+            foreach (var factory in _derivedFactories)
+            {
+                await factory.DisposeAsync().ConfigureAwait(false);
```

Comment:
This should be casted to `IAsyncDisposable` to allow for interface re-implementation on derived classes.

Reply:
done